### PR TITLE
log: Add -timestamp option (cont.)

### PR DESCRIPTION
**Pull request by anlutro**
Continuation of #695 - hope this is OK, this is a feature I'd really like to see added and the original PR seemed abandoned.

Changed the argument parsing to use `sArgs.Token(uIndex)` to allow any order of arguments.

Maybe the same thing is possible with `for (CString& sArg : vsArgs)` but I couldn't see how to grab the element in the next index. Let me know if this is the preferred way to do it.

Also see inline comments.

### CODE REVIEW

File: modules/log.cpp [hunk 1]
```diff
@@ -268,15 +269,35 @@ CString CLogMod::GetServer()
 
 bool CLogMod::OnLoad(const CString& sArgs, CString& sMessage)
 {
+	CString sMyArgs = sArgs;
```
Comment (anlutro) [inline modules/log.cpp]: Not sure if this copying is necessary, it was done this way in awaystore.cpp so I kept it like that.

File: modules/log.cpp [hunk 2]
```diff
@@ -268,15 +269,35 @@ CString CLogMod::GetServer()
...
-	m_sLogPath = sArgs.Token(uIndex);
+	bool bHaveLogPath = false;
+
+	do {
+		sArg = s.sMyArgs.Token(uIndex);
+
+		if (sArg.Equals("-sanitize")) {
+			m_bSanitize = true;
+		} else if (sArg.Equals("-timestamp")) {
+			sArg = s.sMyArgs.Token(++uIndex);
```
Comment (anlutro) [inline modules/log.cpp]: Can this error? If a user loads the module without a timestamp argument, i.e. `LoadModule = log -timestamp`, what will happen?

File: modules/log.cpp [hunk 3]
```diff
@@ -268,15 +269,35 @@ CString CLogMod::GetServer()
...
-	m_sLogPath = sArgs.Token(uIndex);
+	bool bHaveLogPath = false;
+
+	do {
+		sArg = sMyArgs.Token(uIndex);
+
+		if (sArg.Equals("-sanitize")) {
+			m_bSanitize = true;
+		} else if (sArg.Equals("-timestamp")) {
+			sArg = sMyArgs.Token(++uIndex);
```
Comment (anlutro) [inline modules/log.cpp]: Can this error? If a user loads the module without a timestamp argument, i.e. `LoadModule = log -timestamp`, what will happen?

File: modules/log.cpp (hunk 1 above)
Comment (DarthGandalf) [inline modules/log.cpp]: Well, do you need the copy?
Awaystore parses differently, it uses Token(0) every time, and chops the first token out of string every time. So it needs the string to be modifiable.

File: modules/log.cpp (hunk 3 above)
Comment (DarthGandalf) [inline modules/log.cpp]: `""`

File: modules/log.cpp [hunk 4]
```diff
@@ -268,15 +269,35 @@ CString CLogMod::GetServer()
...
-		m_bSanitize = true;
-		++uIndex;
-	}
 
-	// Use load parameter as save path
-	m_sLogPath = sArgs.Token(uIndex);
+	bool bHaveLogPath = false;
+
+	do {
+		sArg = sMyArgs.Token(uIndex);
```
Comment (DarthGandalf) [inline modules/log.cpp]: `.Split()` and iterating through `vector<string>` would be better

File: modules/log.cpp [hunk 5]
```diff
@@ -268,15 +269,35 @@ CString CLogMod::GetServer()
...
+
+		if (sArg.Equals("-sanitize")) {
+			m_bSanitize = true;
+		} else if (sArg.Equals("-timestamp")) {
+			sArg = sMyArgs.Token(++uIndex);
+			m_sTimestamp += sArg + " ";
```
Comment (DarthGandalf) [inline modules/log.cpp]: Does this mean that `-timestamp A -timestamp B` leads to `A B `?

File: modules/log.cpp [hunk 6]
```diff
@@ -313,7 +334,7 @@ bool CLogMod::OnLoad(const CString& sArgs, CString& sMessage)
 		sMessage = "Invalid log path ["+m_sLogPath+"].";
 		return false;
 	} else {
-		sMessage = "Logging to ["+m_sLogPath+"].";
+		sMessage = "Logging to ["+m_sLogPath+"]. Using timestamp ["+m_sTimestamp+"]";
```
Comment (DarthGandalf) [inline modules/log.cpp]: s/timestamp/timestamp format/

**Comment by DarthGandalf** [comment]
Did you test this?

**Comment by DarthGandalf** [comment]
> Maybe the same thing is possible with for (CString& sArg : vsArgs) but I couldn't see how to grab the element in the next index. Let me know if this is the preferred way to do it.

The preferred way would be to have a common way to parse arguments to modules, something like getopt. But ZNC doesn't currently provide such way.

**Comment by jpnurmi** [comment]
I thought the plan was to kill module arguments altogether. Just make them
all configurable options?

**Comment by DarthGandalf** [comment]
> I thought the plan was to kill module arguments altogether.

There are many different plans, and not all of them make sense together :)

**Comment by anlutro** [comment]
Thanks for the comments, I've fixed up all the issues and tested locally. The method used for grabbing the actual timestamp format argument seems kinda primitive but I guess it works.

```
[ .. ] Loading global module [log]...
[ >> ] Logging to [/home/andreas/.znc/moddata/log/$USER/$NETWORK/$WINDOW/%Y-%m-%d.log]. Using timestamp format [%Y-%m-%d %H:%m%:%s] [/usr/local/lib/znc/log.so]
```

**Comment by DarthGandalf** [comment]
@anlutro this pull request doesn't actually change any timestamp...
I'll repeat my question: Did you test this?

**Comment by anlutro** [comment]
Been too long for me to remember exactly what I did. Could be something lost when I was working on the code. In any case, I fixed it up and pushed.

**Comment by henbruas** [comment]
Why is this merged but not included in 1.6.2 nor in the 1.6.3 RCs?

**Comment by DarthGandalf** [comment]
Because it's merged in master, not in 1.6.x.

**Comment by dgw** [comment]
It was merged into master. The 1.6.x releases come from a different branch and receive only minor updates, mostly security fixes.

**Comment by henbruas** [comment]
Oh, I'm assuming that means it won't be included until the next major release? Thanks for the very quick response, guess I'll build from the master branch instead of using 1.6.2.

**Comment by dgw** [comment]
Correct, master will be used for the next major release (1.7). Meanwhile, [building from Git](http://wiki.znc.in/Git) is a good option if you need something that isn't in a release version yet.

**Pull request merged**